_posts/2019-05-21-i18n-messages.adoc

```diff
@@ -0,0 +1,327 @@
+---
+compatible_cuba_versions: 7.0+
+compatible_java_versions: 8+
+project_id: cuba-petclinic-i18n-messages
+permalink: localization-in-cuba-applications
+---
+= Localization in CUBA applications
+:showtitle:
+:sectlinks:
+:sectanchors:
+:page-navtitle: Localization in CUBA applications
+:page-excerpt: In this guide the localization features of a CUBA application are explained
+:page-root: ../../../
+:project_id: cuba-petclinic-i18n-messages
+:java_version: 1.8
+:cuba_version: 7.0
+
+
+CUBA applications have build-in support for localized messages for different languages. Translations are externalized in
+properties files to have a clear separation between the application code and the translated messages.
+
+
+== What Will be Built
+
+This guide enhances the https://github.com/cuba-platform/cuba-petclinic[CUBA petclinic] example to show how localized messages within the CUBA application can be defined and used:
+
+* the application is also available in German and Russian language
```

Comment:
In fact, the app is localized in German and English. Let's drop support for Russian, so it will be easier to support for you.
Also, you haven't applied German translations for platform, so app screens have English messages for standard actions, filter, etc. I think it looks inconsistent with what is stated in the guide.

Reply:
> In fact, the app is localized in German and English. Let's drop support for Russian, so it will be easier to support for you.

My original plan was to ask someone of you to do the translation, but we can also drop it.

> Also, you haven't applied German translations for platform, so app screens have English messages for standard actions, filter, etc. I think it looks inconsistent with what is stated in the guide.

Yes, this is right. What I did is to create translations for the domain model of the app and the custom screen translations. I'm not sure which route is better to go:

On the one side what we see there is exactly something that a non-engish / non-russian application looks like after it was created with translations for the app specific stuff. So in this regard, the guide takes them step by step and arrives at the final stage, where the community translations are applied.

On the other hand, it looks not very fancy - you are right. There is one image showing the editor in german locale, where everything is german except for "OK" / "Cancel" buttons.


Why I'm not really sure how to do it, is because down below in the `Manual Providing Translations` section, the whole point is to show manual translations. But if we just use the app component for DE locale,  this is not really needed. But indeed in case the language is not available out of the box, it is very important information to have.


How about we do it the following way:

1. we will provide russian translations for the domain model
2. we switch from the main language in this guide from german to russian
3. In the section `Translation of Existing UI Screens / Functionality` we switch to german language
4. In `Community Language Packs` we pick the german app component for translations
5. In `Manual Providing Translations` we show how to fix missing / wrong translations (as they exists in german translation app component probably)

This way we have a consistent experience throughout the screenshots.

Reply:
I see your point and agree that the guide should show the process step by step. In this case we just make some remarks for what the user sees at the current step, why it is, and how it will be improved later in the guide.

Regarding what to show, let's try to imagine who is interested is this guide the most. In my opinion, they are developers of apps for local non-anglophone countries. And Germany is a perfect example. To keep things simple, one non-English language is enough, so I don't think we need Russian. Moreover, Russian is not a typical use case because the framework itself contains Russian localization files.

So from the practical purpose point of view, the scenario could be as follows:

1. Make German default language, add localization for app screens and entities. Show how it looks in the app.

2. Try to localize some common elements as in your `Manually Provided Translation`. Show how it looks in the app. This step is useful for those who don't have community translations.

3. Use a whole translation from community repo or add-on.
 
What do you think?

Reply:
> Make German default language

Do you mean making it use the `messages.properties` file? I'm not sure how this would work, since the addon uses the `_de` files (see: https://github.com/balvi/translation-de/blob/master/modules/global/src/com/haulmont/cuba/messages_de.properties)

Besides that, the approach is fine for me 👍

Reply:
> Do you mean making it use the messages.properties file?
Yes. It should work because the algorithm of searching for message is simple: first look for the key in `messages_de` files, then if not found in `messages`. So the app's files don't have any relation to add-on files, they can be differently organized.